esphome/components/nabu/flac_decoder.cpp

```diff
@@ -355,51 +490,93 @@ FLACDecoderResult FLACDecoder::decode_residuals(uint32_t block_size) {
     return FLAC_DECODER_ERROR_BLOCK_SIZE_NOT_DIVISIBLE_RICE;
   }
 
-  for (std::size_t i = 0; i < num_partitions; i++) {
-    uint32_t count = block_size >> partition_order;
-    if (i == 0) {
-      count -= this->block_result_.size();
-    }
+  int32_t *out_ptr = sub_frame_buffer + warm_up_samples;
+  {
+    uint32_t count = (block_size >> partition_order) - warm_up_samples;
     uint32_t param = this->read_uint(param_bits);
     if (param < escape_param) {
       for (std::size_t j = 0; j < count; j++) {
-        this->block_result_.push_back(this->read_rice_sint(param));
+        *(out_ptr++) = this->read_rice_sint(param);
       }
     } else {
       std::size_t num_bits = this->read_uint(5);
+      if( num_bits == 0 ){
+        std::memset( out_ptr, 0, count * sizeof(int32_t));
+        out_ptr += count;
+      } else {
+        for (std::size_t j = 0; j < count; j++) {
+          *(out_ptr++) = this->read_sint(num_bits);
+        }
+      }
+    }
+  }
+
+  uint32_t count = block_size >> partition_order;
+  for (std::size_t i = 1; i < num_partitions; i++) {
+    uint32_t param = this->read_uint(param_bits);
+    if (param < escape_param) {
       for (std::size_t j = 0; j < count; j++) {
-        this->block_result_.push_back(this->read_sint(num_bits));
+        *(out_ptr++) = this->read_rice_sint(param);
+      }
+    } else {
+      std::size_t num_bits = this->read_uint(5);
+      if( num_bits == 0 ){
+        std::memset( out_ptr, 0, count * sizeof(int32_t));
+        out_ptr += count;
+      } else {
+        for (std::size_t j = 0; j < count; j++) {
+          *(out_ptr++) = this->read_sint(num_bits);
+        }
       }
     }
   }  // for each partition
 
   return FLAC_DECODER_SUCCESS;
 }  // decode_residuals
 
-void FLACDecoder::restore_linear_prediction(const std::vector<int16_t> &coefs, int32_t shift) {
-  for (std::size_t i = 0; i < this->block_result_.size() - coefs.size() + 1; i++) {
+void FLACDecoder::restore_linear_prediction(int32_t* sub_frame_buffer, size_t num_of_samples, const std::vector<int16_t> &coefs, int32_t shift) {
+  
+  for (std::size_t i = 0; i < num_of_samples - coefs.size() + 1; i++) {
     int32_t sum = 0;
     for (std::size_t j = 0; j < coefs.size(); ++j) {
-      sum += (this->block_result_[i + j] * coefs[j]);
+      sum += (sub_frame_buffer[i + j] * coefs[j]);
     }
-    this->block_result_[i + coefs.size() - 1] = (sum >> shift);
+    sub_frame_buffer[i + coefs.size() - 1] = (sum >> shift);
   }
 }  // restore_linear_prediction
 
-uint32_t FLACDecoder::read_uint(std::size_t num_bits) {
+uint32_t FLACDecoder::read_aligned_byte(){
+  //assumes byte alignment
+  assert( this->bit_buffer_length_ % 8 == 0 );
```

Comment:
Can we fail more gracefully and avoid using ``assert`` here?

Reply:
Actually, the assertion case shouldn't be triggered due to errors in parsing the bitstream, but only if the function is used in an unintended way by a developer. So, if you prefer I can simply remove that line.